src/PaymentServer/Processors/Stripe.hs

```diff
@@ -83,3 +119,49 @@ webhook d Event{eventId=Just (EventId eventId), eventType=ChargeSucceededEvent,
 webhook d _ =
   -- TODO: Record the eventId somewhere.
   return Ok
+
+
+-- | Browser facing API that takes token, voucher and a few other information
+-- and calls stripe charges API. If payment succeeds, then the voucher is stored
+-- in the voucher database.
+type ChargesAPI = "charge" :> ReqBody '[JSON] Charges :> Post '[JSON] Acknowledgement
+
+data Charges = Charges
+  { token :: Text
+  , voucher :: Voucher
+  , amount :: Int
+  , currency :: Text
+  } deriving (Show, Eq)
+
+instance FromJSON Charges where
+  parseJSON (Object v) = Charges <$>
+                         v .: "token" <*>
+                         v .: "voucher" <*>
+                         v .: "amount" <*>
+                         v .: "currency"
+  parseJSON _ = mzero
+
+-- | call the stripe Charge API (with token, voucher in metadata, amount, currency etc
+-- and if the Charge is okay, then set the voucher as "paid" in the database.
+charge :: VoucherDatabase d => d -> ByteString -> Charges -> Handler Acknowledgement
+charge d key (Charges token voucher amount currency) = do
+  let config = StripeConfig (StripeKey key) Nothing
+      tokenId = TokenId token
+  result <- liftIO $ stripe config $
+    createCharge (Amount amount) (read (unpack currency))
```

Comment:
is `read` total?  I guess if it fails this will turn into an error response somehow because Servant probably handles IO exceptions?  We might want to handle these and turn them into "bad request" instead of (I'm guessing) "internal server error" someday.  For now, doesn't matter - I think we may even hard-code USD into the browser form to start with.

Reply:
I had made a note to change this, but it slipped (along with other things) before I sent the PR. I just pushed a commit.